File: genium/src/main/java/com/here/genium/model/cpp/CppMethod.java
/*
 * Copyright (C) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.model.cpp;

import java.util.EnumSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;
import lombok.EqualsAndHashCode;
import lombok.Singular;

@EqualsAndHashCode(callSuper = true)
public final class CppMethod extends CppElementWithComment {

  public final CppTypeRef returnType;
  public final String returnComment;
  public final Set<Specifier> specifiers;
  public final Set<Qualifier> qualifiers;
  public final List<CppParameter> parameters;

  public enum Specifier {
    EXPLICIT("explicit"),
    INLINE("inline"),
    STATIC("static"),
    VIRTUAL("virtual");

    private final String text;

    Specifier(final String text) {
      this.text = text;
    }

    @Override
    public String toString() {
      return text;
    }
  }

  public enum Qualifier {
    CONST("const"),
    OVERRIDE("override"),
    PURE_VIRTUAL("= 0");

    private final String text;

    Qualifier(final String text) {
      this.text = text;
    }

    @Override
    public String toString() {
      return text;
    }
  }

  @SuppressWarnings("ParameterNumber")
  @lombok.Builder(builderClassName = "Builder")
  private CppMethod(
      final String name,
      final String comment,
      final CppTypeRef returnType,
      final String returnComment,
      final String fullyQualifiedName,
      @Singular final Set<Specifier> specifiers,
      @Singular final Set<Qualifier> qualifiers,
      @Singular final List<CppParameter> parameters) {
    super(name, fullyQualifiedName == null ? name : fullyQualifiedName, comment);
    this.returnType = returnType != null ? returnType : CppPrimitiveTypeRef.VOID;
    this.returnComment = returnComment;
    this.specifiers =
        !specifiers.isEmpty() ? EnumSet.copyOf(specifiers) : EnumSet.noneOf(Specifier.class);
    this.qualifiers =
        !qualifiers.isEmpty() ? EnumSet.copyOf(qualifiers) : EnumSet.noneOf(Qualifier.class);
    this.parameters = new LinkedList<>(parameters);
  }

  @SuppressWarnings("unused")
  public static class Builder {
    private String name;

    Builder() {
      this(null);
    }

    public Builder(final String name) {
      this.name = name;
    }
  }

  @Override
  public Stream<? extends CppElement> stream() {
    return Stream.concat(Stream.of(returnType), parameters.stream());
  }
}

File: genium/src/main/java/com/here/genium/model/cpp/CppMethod.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.model.cpp

import com.here.genium.model.common.Comments
import java.util.EnumSet
import java.util.stream.Stream

class CppMethod(
    name: String,
    fullyQualifiedName: String = name,
    comment: Comments = Comments(),
    val returnType: CppTypeRef = CppPrimitiveTypeRef.VOID,
    val returnComment: String? = null,
    val errorEnumName: String? = null,
    val isNotNull: Boolean = false,
    val parameters: List<CppParameter> = emptyList(),
    val specifiers: Set<Specifier> = EnumSet.noneOf(Specifier::class.java),
    val qualifiers: Set<Qualifier> = EnumSet.noneOf(Qualifier::class.java)
) : CppElementWithComment(name, fullyQualifiedName, comment = comment) {

    enum class Specifier(private val text: String) {
        EXPLICIT("explicit"),
        INLINE("inline"),
        STATIC("static"),
        VIRTUAL("virtual");

        override fun toString() = text
    }

    enum class Qualifier(private val text: String) {
        CONST("const"),
        OVERRIDE("override"),
        PURE_VIRTUAL("= 0");

        override fun toString() = text
    }

    @Suppress("unused")
    fun hasComment() = !comment.isEmpty ||
        hasReturnComment() ||
        parameters.stream().anyMatch(CppParameter::hasComment)

    private fun hasReturnComment() = !returnComment.isNullOrEmpty() || isNotNull

    @Suppress("unused")
    val isReturningVoid = returnType == CppPrimitiveTypeRef.VOID

    override fun stream() = Stream.concat(Stream.of(returnType), parameters.stream())

    fun copy(specifiers: Set<Specifier>? = null, qualifiers: Set<Qualifier>? = null) =
        CppMethod(
            name,
            fullyQualifiedName,
            comment,
            returnType,
            returnComment,
            errorEnumName,
            isNotNull,
            parameters,
            specifiers ?: this.specifiers,
            qualifiers ?: this.qualifiers
        )
}
